### ISSUE TITLE
Shiftit doesn't work with two monitors if one is in Fullscreen mode

### ISSUE THREAD

**Issue by dylnclrk** [fikovnik/ShiftIt#274]
Problem:
=======
I can't use Shiftit when one of my monitors is in full-screen mode.

More detail:
=========

I have two monitors set up today. In one monitor I have Terminal.app running in full-screen mode, in the other I have several applications (including Chrome) running in regular, windowed mode.

When trying to resize the focused Chrome window, nothing happens, and I see,

```
2018-02-26 13:27:23.813 ShiftIt[31982/0x7fffb50e1340] [lvl=3] -[ShiftItAppDelegate invokeShiftItActionByIdentifier_:] Execution of ShiftIt action: right failed: Windows in fullscreen are not supported
NSError stack trace:
org.shiftitapp.shifit.error:20103 - Windows in fullscreen are not supported
```

in the debug log.

So it seems like it's trying to resize my full-screened Terminal window.

Workaround: 
=========

Take Terminal out of full-screen mode, and everybody's happy. Shiftit works.

**Comment by fikovnik** [comment]
Which version do you use - try the latest from homebrew - should work.

**Comment by dylnclrk** [comment]
@fikovnik I tried updating (via homebrew cask) before posting this bug report. Unfortunately I'm still seeing the issue on 1.6.5 (and Mac OS 10.13.3).

Oddly enough, I also noticed an inconsistency in the reported version number… see #275 when I clicked "Check for updates". So maybe I'm messing something up with the update?

I wish I were more handy with OS X so I could do a better job helping with these issues. 

Let me know if there's anything else I can do to help triage.

**Comment by nikcorg** [comment]
Ditto for this issue on OS X 10.12.6 and ShiftIt 1.6.6.

**Comment by emptylambda** [comment]
Same issue with OSX 10.13 and ShiftIt 1.6.3

**Comment by dylnclrk** [comment]
I updated to Shiftit 1.6.6, and I'm still seeing this issue. As @emptylambda says it seems to be somewhat inconsistent and depend on the application that's full-screened.

### Some scenarios

**Fullscreened application**: Safari
**Windowed application on other screen**: Chrome/Terminal.app/Spotify/Safari
**Behavior**: Works as normal

**Fullscreen**: Terminal.app
**Windowed**: Chrome/Spotify/Safari
**Behavior**: Doesn't work. But if there is a terminal.app window in the background of the windowed screen, it gets resized (even if you have focus in another application).

**Fullscreen**: Chrome
**Windowedn**: Spotify/Terminal.app/Safari
**Behavior**: Doesn't work. But if there is a Chrome window in the background of the windowed screen, it gets resized (even if you have focus in another application).

**Comment by fosron** [comment]
Getting this with 3 monitors (1 - Slack in fullscreen, 2 - App i want to maximize with ShiftIt, 3 - PHPStorm);

macOS 10.13.2 + ShiftIt 1.6.6

```
2018-03-05 10:39:56.791 ShiftIt[20848/0x7fff88491340] [lvl=3] -[ShiftItAppDelegate invokeShiftItActionByIdentifier_:] Execution of ShiftIt action: maximize failed: Windows in fullscreen are not supported
NSError stack trace:
org.shiftitapp.shifit.error:20103 - Windows in fullscreen are not supported
```

**Comment by emptylambda** [comment]
Im looking into the codebase for this, it seems like Terminal-like app automagically obtains the highest priority for CFWindows list and will always get chosen as if it is "focused" windows. 
This is located in `SIWindowManager.m` the function `getFocusedWindow` (around line 144) 
Since Im not familiar with Core, my fix may take a bit longer but Im trying to mend this :)

**Comment by nikcorg** [comment]
I can repeat the behaviour with any full screen app, not just terminal (iTerm2). 

I switched to Spectacle.app which doesn't have this problem.

**Comment by aguynamedben** [comment]
This is happening to me too. MacOS 10.13.3, ShiftIt 1.6.6. MacBook Pro with 4k Dell monitor hooked up. Trying to work on the 4k monitor and watch The Masters on the MacBook Pro display. 🖥⛳️

Thanks to the ShiftIt team for the hard work developing this free app. 👏

**Comment by nikcorg** [comment]
> Thanks to the ShiftIt team for the hard work developing this free app. 👏

Seconded. I failed to mention in my previous comment that my switch was reluctant. Apart from being a great app otherwise, ShiftIt has a universal shortcut to fullscreen an app on OSX, which for some strange reason doesn't have a standard OS level shortcut.

**Comment by UrGuardian4ngel** [comment]
> ShiftIt has a universal shortcut to fullscreen an app on OSX, **which for some strange reason doesn't have a standard OS level shortcut.**

@nikcorg Doesn't it? Well, maybe not a _standard_ one.  
But, nothing's stopping you from customizing your system, and creating one yourself?

Seems like you can simply create your own custom shortcut for that. As a matter of fact, you can create your own shortcuts for **any** macOS app (separately), just by adding a _menu item_'s title, and hitting a shortcut?

System Preferences > Keyboard > Shortcuts

<img width="780" alt="image" src="https://user-images.githubusercontent.com/3755487/38713352-0a2386ac-3ed1-11e8-9f16-28d1c72ca426.png">

**Comment by nikcorg** [comment]
@UrGuardian4ngel yes, I've set that keyboard shortcut, but despite that all apps don't adhere to it. E.g. iTerm listens to option-enter, VS Code listens to ctrl-option-f.

**Comment by UrGuardian4ngel** [comment]
Hmm... For me, both seem to work?
- my "custom" `cmd + ctrl + alt + F` keybinding
- _and_ the "application default", like `cmd + enter` for e.g. iTerm

I'm on macOS 10.13.3.

**Comment by hadidotj** [comment]
FYI, it appears PR #276 fixes this issue, for those who wish to build locally.

**Comment by elijahcarrel** [comment]
Note that for those of you who are not mac or iOS developers but are trying to apply the patch, the Xcode tools need to be installed _and Xcode needs to have been opened at least once_ in order for the `xcodebuild` statement that @hadidotj mentioned above to work. If you (like me) downloaded Xcode to run that statement but did not open it, you get the unhelpful error message "Command /Applications/Xcode.app/Contents/Developer/usr/bin/ibtool failed with exit code 6. The tool may have crashed."

**Comment by DannyDannyDanny** [comment]
### Solution for `tool 'xcodebuild' requires Xcode`
When running `xcodebuild -target "ShiftIt NoX11" -configuration Release
` you may get the error:
```
tool 'xcodebuild' requires Xcode, but active developer directory '/Library/Developer/CommandLineTools' is a command line tools instance
```
To get around this:
1. Install **Xcode from App Store**
2. Open Xcode, finish install, **accept license**
3. Point Xcode-select to the right directory `sudo xcode-select -s /Applications/Xcode.app/Contents/Developer`
4. Go back to following instruction by @hadidotj 
***
Credit to this [stackoverflow thread](https://stackoverflow.com/questions/17980759/xcode-select-active-developer-directory-error/17980786#17980786)

**Comment by holmesmi** [comment]
Question: if I have multiple monitors, and some windows in full-screen mode, should I be able to use the "Next" and "Previous" shortcuts, or are they intended to only work on windowed apps? I don't know if I should file a feature request or a bug report....

**Comment by jaggerwang** [comment]
Came into the same problem when open a fullscreen Chrome window on my second monitor, it became ok when exit fullscreen mode. And it is also ok to open a fullscreen Chrome window on my first monitor.

```
2018-09-04 10:19:28.355 ShiftIt[5569/0x7fff97844380] [lvl=3] -[ShiftItAppDelegate invokeShiftItActionByIdentifier_:] Execution of ShiftIt action: maximize failed: Windows in fullscreen are not supported
NSError stack trace:
org.shiftitapp.shifit.error:20103 - Windows in fullscreen are not supported
2018-09-04 10:19:48.602 ShiftIt[5569/0x7fff97844380] [lvl=3] -[ShiftItAppDelegate invokeShiftItActionByIdentifier_:] Execution of ShiftIt action: maximize failed: Windows in fullscreen are not supported
NSError stack trace:
org.shiftitapp.shifit.error:20103 - Windows in fullscreen are not supported
```

**Comment by MladenJanjetovic** [comment]
Same here, I had this issue but didn't realize that it has something to do with Full screen mode on one of the screens. Thanks

**Comment by L33t-HAXXOR** [comment]
Same here for me running 10.14.6 with ShiftIt 1.6.6
![System - Mac Config 2020-12-01](https://user-images.githubusercontent.com/31870368/132073081-42ec2a15-4474-402b-9a44-4211e78ecfd5.png)
![image](https://user-images.githubusercontent.com/31870368/132073131-1960a68f-aba4-4c84-adfb-4d1fd166bac3.png)

**Note**
It appears this issue is a duplicate of a [previous issue](https://github.com/fikovnik/ShiftIt/issues/249)

**Comment by holmesmi** [comment]
> Same here for me running 10.14.6 with ShiftIt 1.6.6
> ![System - Mac Config 2020-12-01](https://user-images.githubusercontent.com/31870368/132073081-42ec2a15-4474-402b-9a44-4211e78ecfd5.png)
> ![image](https://user-images.githubusercontent.com/31870368/132073131-1960a68f-aba4-4c84-adfb-4d1fd166bac3.png)
> 
> **Note**
> It appears this issue is a duplicate of a [previous issue](https://github.com/fikovnik/ShiftIt/issues/249)

It appears ShiftIt has been abandoned. I'm using https://rectangleapp.com/ for the same purpose, and it works great.
